commons-db/src/main/java/com/palantir/nexus/db/ResourceCreationLocation.java

```diff
@@ -0,0 +1,99 @@
+/**
+ * Copyright 2015 Palantir Technologies
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.nexus.db;
+
+import java.io.Serializable;
+
+import org.joda.time.format.ISODateTimeFormat;
+
+import com.google.common.base.MoreObjects;
+
+/**
+ * A RuntimeException that represents the creation location of a resource (e.g. JDBC Connection or
+ * ResultSet) to help identify potential resource leaks. This instance tracks the stack trace of
+ * resource creation as well as the thread name and ID of the creator.
+ */
+public class ResourceCreationLocation extends RuntimeException {
```

Comment:
Been meaning to clean this up, probably as a followup change, but should move the creation timestamp to the ResourceCreationLocation and have ThreadInfo implement equals/hashCode

```
public class ResourceCreationLocation extends RuntimeException {
    private static final long serialVersionUID = 1L;

    public static final long UNSET_THREAD_ID = -1L;

    private final ThreadInfo creatingThreadInfo;
    private final long creationTimestamp;

    public ResourceCreationLocation(String message) {
        this(message, null);
    }

    public ResourceCreationLocation(String message, Exception cause) {
        super(message, cause);
        this.creatingThreadInfo = ThreadInfo.from(Thread.currentThread());
        this.creationTimestamp = System.currentTimeMillis();
    }

    public ThreadInfo getCreatingThreadInfo() {
        return creatingThreadInfo;
    }

    public long getCreationTimestamp() {
        return creationTimestamp;
    }

    public String formattedCreationTimestamp() {
        return ISODateTimeFormat.dateTime().print(creationTimestamp);
    }

    @Override
    public String getMessage() {
        return super.getMessage()
                + " at " + formattedCreationTimestamp()
                + " by " + creatingThreadInfo;
    }

    /**
     * Used to avoid strong references to {@link Thread}
     */
    public static class ThreadInfo implements Serializable {
        private static final long serialVersionUID = 1L;

        private final String name;
        private final long id;

        public static ThreadInfo from(Thread t) {
            return new ThreadInfo(t.getName(), t.getId());
        }

        private ThreadInfo(String name, long id) {
            super();
            this.name = name;
            this.id = id;
        }

        public String getName() {
            return name;
        }

        public long getId() {
            return id;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) {
                return true;
            }
            if (o == null || getClass() != o.getClass()) {
                return false;
            }
            ThreadInfo that = (ThreadInfo) o;
            return id == that.id && Objects.equal(name, that.name);
        }

        @Override
        public int hashCode() {
            int result = name == null ? 0 : name.hashCode();
            result = 31 * result + (int) (id ^ (id >>> 32));
            return result;
        }

        @Override
        public String toString() {
            return MoreObjects.toStringHelper("thread")
                    .add("name", name)
                    .add("ID", id)
                    .toString();
        }
    }

}
```